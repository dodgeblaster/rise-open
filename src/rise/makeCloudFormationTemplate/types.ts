export type ResolverInstructions = {
    type: SchemaType
    field: string
    vtlRequest: string
    vtlResponse: string
}

export type EventTemplateInstructions = {
    type: 'Events'
    field: string
    inputTemplate: string
}

export type SchemaType = 'Query' | 'Mutation'

export type ResourceDefinition = {
    Resources: object
    Outputs: object
}

export type CloudformationTemplate = {
    Resources: object
    Outputs: object
}
