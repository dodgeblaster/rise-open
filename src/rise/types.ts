/**
 * RiseCommandInput
 * This represents the input from the cli deploy command.
 *
 */
export interface RiseCommandInput {
    profile?: string
    region?: string
    stage?: string
}

/**
 * Rise Block
 * This represents the entire project definition and will inform
 * how the cloudformation should be built, and how appsync should
 * be configured
 *
 */
export interface RiseBlock {
    api: RiseBlockApi
    code: RiseBlockCode
    config: RiseBlockConfig
}

type RiseBlockConfig = {
    name: string
    profile: string
    region: string
    stage: string
    auth: boolean
    env: Record<string, string>
    s3BucketName: string
    s3BucketFile: string
    events: string[]
    additionalUserPool?: string
}

type RiseBlockApi = string
type RiseBlockCode = {
    Query?: {
        [name: string]: Function | Resolver
    }
    Mutation?: {
        [name: string]: Function | Resolver
    }
    Events?: {
        [name: string]: Function | Resolver
    }
}

//ny // need to make this into a proper type

/**
 * Rise Output
 * Once the app has been deployed, the following details we be outputed
 * which will allow us to print info to the terminal, and write a helpful
 * info.md file. In the future, this may aslo be where we derive state.
 *
 */
export interface RiseOutput {
    endpoint: string
    table: string
    apiKey?: string
    userPool?: UserPoolSettings
}

type UserPoolSettings = {
    userPoolId: string
    userPoolRegion: string
    userPoolClientId: string
}

/**
 * Actions
 *
 */
type Action = 'db' | 'add' | 'guard'

type DbActionDefinition = {
    type: 'db'
    action: 'create' | 'remove' | 'get' | 'list'
    sk?: string
    pk?: string
    pk2?: string
    pk3?: string

    // retry: {

    // },

    // catch: {

    // }
}

type AddActionDefinition = {
    type: 'add'
    [value: string]: string
}

type GuardActionDefinition = {
    type: 'guard'
    inputPath?: 'string' // if schema has input, then we say take everything from input, rather than root
    outputMerge?: 'string' // think thru
    outputPath?: 'string' // think thru

    sk: string
    pk?: string
    pk2?: string
    pk3?: string
}

type FunctionActionDefinition = {
    type: 'function'
}

type ActionDefinition =
    | DbActionDefinition
    | AddActionDefinition
    | GuardActionDefinition
    | FunctionActionDefinition

export type Resolver = ActionDefinition | ActionDefinition[]
