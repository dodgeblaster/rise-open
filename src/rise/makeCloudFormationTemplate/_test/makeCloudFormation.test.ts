import build from '../index'
import { RiseBlock } from '../../types'

test('build resolver instructions works', () => {
    const block: RiseBlock = {
        api: ``,
        code: {
            Mutation: {
                submit: [
                    {
                        type: 'guard',
                        pk: '123',
                        sk: '123'
                    },
                    {
                        type: 'add',
                        note: '!email'
                    },
                    {
                        type: 'function'
                    },
                    {
                        type: 'db',
                        action: 'create'
                    }
                ],
                check: {
                    type: 'guard',
                    pk: '123',
                    sk: '123'
                },

                add: {
                    type: 'add',
                    note: '!email'
                },
                hat: {
                    type: 'function'
                },
                note: {
                    type: 'db',
                    action: 'create'
                }
            },
            Query: {
                submit: [
                    {
                        type: 'guard',
                        pk: '123',
                        sk: '123'
                    },
                    {
                        type: 'add',
                        note: '!email'
                    },
                    {
                        type: 'function'
                    },
                    {
                        type: 'db',
                        action: 'create'
                    }
                ],
                check: {
                    type: 'guard',
                    pk: '123',
                    sk: '123'
                },

                add: {
                    type: 'add',
                    note: '!email'
                },
                hat: {
                    type: 'function'
                },
                note: {
                    type: 'db',
                    action: 'create'
                }
            },
            Events: {}
        },
        config: {
            auth: false,
            env: {},
            name: 'hi',
            profile: 'example',
            region: 'us-east-2',
            s3BucketFile: 'rise-lambdadeployments',
            s3BucketName: 'rise-hi-dev.zip',
            stage: 'dev',
            events: []
        }
    }

    const result = build(block)
    // console.log('+++ ', JSON.stringify(result, null, ' '))
    expect(result).toMatchSnapshot()
})
