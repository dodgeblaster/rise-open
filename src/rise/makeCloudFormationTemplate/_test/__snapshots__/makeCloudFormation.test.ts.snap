// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build resolver instructions works 1`] = `"{\\"AWSTemplateFormatVersion\\":\\"2010-09-09\\",\\"Description\\":\\"AWS CloudFormation Rise Template\\",\\"Resources\\":{\\"GraphQlDslambdadatasourceRole\\":{\\"Type\\":\\"AWS::IAM::Role\\",\\"Properties\\":{\\"AssumeRolePolicyDocument\\":{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":[\\"appsync.amazonaws.com\\"]},\\"Action\\":[\\"sts:AssumeRole\\"]}]},\\"Policies\\":[{\\"PolicyName\\":\\"GraphQlDslambdadatasourcePolicy\\",\\"PolicyDocument\\":{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":[\\"lambda:invokeFunction\\"],\\"Effect\\":\\"Allow\\",\\"Resource\\":[{\\"Fn::GetAtt\\":[\\"GraphQLEndpointLambdaFunction\\",\\"Arn\\"]},{\\"Fn::Join\\":[\\":\\",[{\\"Fn::GetAtt\\":[\\"GraphQLEndpointLambdaFunction\\",\\"Arn\\"]},\\"*\\"]]}]}]}}]}},\\"IamRoleLambdaExecution\\":{\\"Type\\":\\"AWS::IAM::Role\\",\\"Properties\\":{\\"AssumeRolePolicyDocument\\":{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":[\\"lambda.amazonaws.com\\"]},\\"Action\\":[\\"sts:AssumeRole\\"]}]},\\"Policies\\":[{\\"PolicyName\\":{\\"Fn::Join\\":[\\"-\\",[\\"hi-dev\\",\\"lambda\\"]]},\\"PolicyDocument\\":{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Effect\\":\\"Allow\\",\\"Action\\":[\\"logs:CreateLogStream\\",\\"logs:CreateLogGroup\\"],\\"Resource\\":[{\\"Fn::Sub\\":\\"arn:\${AWS::Partition}:logs:\${AWS::Region}:\${AWS::AccountId}:log-group:/aws/lambda/hi-dev*:*\\"}]},{\\"Effect\\":\\"Allow\\",\\"Action\\":[\\"logs:PutLogEvents\\"],\\"Resource\\":[{\\"Fn::Sub\\":\\"arn:\${AWS::Partition}:logs:\${AWS::Region}:\${AWS::AccountId}:log-group:/aws/lambda/hi-dev*:*\\"}]},{\\"Effect\\":\\"Allow\\",\\"Action\\":[\\"dynamodb:Query\\",\\"dynamodb:Scan\\",\\"dynamodb:GetItem\\",\\"dynamodb:PutItem\\",\\"dynamodb:UpdateItem\\",\\"dynamodb:DeleteItem\\"],\\"Resource\\":{\\"Fn::Sub\\":\\"arn:aws:dynamodb:\${AWS::Region}:*:*\\"}},{\\"Effect\\":\\"Allow\\",\\"Action\\":[\\"events:PutEvents\\"],\\"Resource\\":{\\"Fn::GetAtt\\":[\\"EventBridge\\",\\"Arn\\"]}},{\\"Effect\\":\\"Allow\\",\\"Action\\":[\\"cognito-idp:*\\"],\\"Resource\\":{\\"Fn::GetAtt\\":[\\"CognitoUserPoolMyUserPool\\",\\"Arn\\"]}}]}}],\\"Path\\":\\"/\\",\\"RoleName\\":{\\"Fn::Join\\":[\\"-\\",[\\"hi-dev\\",{\\"Ref\\":\\"AWS::Region\\"},\\"lambdaRole\\"]]}}},\\"Database\\":{\\"Type\\":\\"AWS::DynamoDB::Table\\",\\"Properties\\":{\\"TableName\\":\\"hi-dev\\",\\"AttributeDefinitions\\":[{\\"AttributeName\\":\\"pk\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"sk\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"pk2\\",\\"AttributeType\\":\\"S\\"},{\\"AttributeName\\":\\"pk3\\",\\"AttributeType\\":\\"S\\"}],\\"KeySchema\\":[{\\"AttributeName\\":\\"pk\\",\\"KeyType\\":\\"HASH\\"},{\\"AttributeName\\":\\"sk\\",\\"KeyType\\":\\"RANGE\\"}],\\"GlobalSecondaryIndexes\\":[{\\"IndexName\\":\\"pk2\\",\\"KeySchema\\":[{\\"AttributeName\\":\\"pk2\\",\\"KeyType\\":\\"HASH\\"},{\\"AttributeName\\":\\"sk\\",\\"KeyType\\":\\"RANGE\\"}],\\"Projection\\":{\\"ProjectionType\\":\\"ALL\\"}},{\\"IndexName\\":\\"pk3\\",\\"KeySchema\\":[{\\"AttributeName\\":\\"pk3\\",\\"KeyType\\":\\"HASH\\"},{\\"AttributeName\\":\\"sk\\",\\"KeyType\\":\\"RANGE\\"}],\\"Projection\\":{\\"ProjectionType\\":\\"ALL\\"}}],\\"BillingMode\\":\\"PAY_PER_REQUEST\\"}},\\"EventBridge\\":{\\"Type\\":\\"AWS::Events::EventBus\\",\\"Properties\\":{\\"Name\\":\\"hi-dev\\"}},\\"GraphQLEndpointLambdaFunction\\":{\\"Type\\":\\"AWS::Lambda::Function\\",\\"Properties\\":{\\"Code\\":{\\"S3Bucket\\":\\"rise-hi-dev.zip\\",\\"S3Key\\":\\"rise-lambdadeployments\\"},\\"Description\\":\\"\\",\\"Environment\\":{\\"Variables\\":{}},\\"FunctionName\\":\\"hi-dev\\",\\"Handler\\":\\"_index.handler\\",\\"MemorySize\\":256,\\"Role\\":{\\"Fn::GetAtt\\":[\\"IamRoleLambdaExecution\\",\\"Arn\\"]},\\"Runtime\\":\\"nodejs12.x\\",\\"Timeout\\":15},\\"DependsOn\\":[\\"MainFunctionLogGroup\\"]},\\"MainFunctionLogGroup\\":{\\"Type\\":\\"AWS::Logs::LogGroup\\",\\"Properties\\":{\\"LogGroupName\\":\\"/aws/lambda/hi-dev\\"}},\\"GraphQlApi\\":{\\"Type\\":\\"AWS::AppSync::GraphQLApi\\",\\"Properties\\":{\\"AuthenticationType\\":\\"API_KEY\\",\\"Name\\":\\"hi-dev\\",\\"XrayEnabled\\":false}},\\"GraphQlSchema\\":{\\"Type\\":\\"AWS::AppSync::GraphQLSchema\\",\\"Properties\\":{\\"Definition\\":\\"\\",\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]}}},\\"GraphQlApiKeyDefault\\":{\\"Type\\":\\"AWS::AppSync::ApiKey\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"Expires\\":1627859769}},\\"GraphQlDslambdadatasource\\":{\\"Type\\":\\"AWS::AppSync::DataSource\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"Name\\":\\"lambdadatasource\\",\\"Description\\":\\"Main Lambda function\\",\\"Type\\":\\"AWS_LAMBDA\\",\\"ServiceRoleArn\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasourceRole\\",\\"Arn\\"]},\\"LambdaConfig\\":{\\"LambdaFunctionArn\\":{\\"Fn::GetAtt\\":[\\"GraphQLEndpointLambdaFunction\\",\\"Arn\\"]}}}},\\"ResolverQuerysubmit\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Query\\",\\"FieldName\\":\\"submit\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": [{ \\\\n    \\\\\\"type\\\\\\": \\\\\\"guard\\\\\\",\\\\n    \\\\\\"pk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"sk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n},{\\\\n    \\\\\\"type\\\\\\": \\\\\\"add\\\\\\",\\\\n    \\\\\\"note\\\\\\":\\\\\\"!email\\\\\\"\\\\n},{ \\\\n    \\\\\\"type\\\\\\": \\\\\\"function\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\",\\\\n    \\\\\\"userpoolId\\\\\\":\\\\\\"\${UserPoolId}\\\\\\"\\\\n},{ \\\\n    \\\\\\"type\\\\\\": \\\\\\"db\\\\\\",\\\\n    \\\\\\"action\\\\\\": \\\\\\"create\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n}]\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverQuerycheck\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Query\\",\\"FieldName\\":\\"check\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": { \\\\n    \\\\\\"type\\\\\\": \\\\\\"guard\\\\\\",\\\\n    \\\\\\"pk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"sk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverQueryadd\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Query\\",\\"FieldName\\":\\"add\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": {\\\\n    \\\\\\"type\\\\\\": \\\\\\"add\\\\\\",\\\\n    \\\\\\"note\\\\\\":\\\\\\"!email\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverQueryhat\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Query\\",\\"FieldName\\":\\"hat\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": { \\\\n    \\\\\\"type\\\\\\": \\\\\\"function\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\",\\\\n    \\\\\\"userpoolId\\\\\\":\\\\\\"\${UserPoolId}\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverQuerynote\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Query\\",\\"FieldName\\":\\"note\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": { \\\\n    \\\\\\"type\\\\\\": \\\\\\"db\\\\\\",\\\\n    \\\\\\"action\\\\\\": \\\\\\"create\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverMutationsubmit\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Mutation\\",\\"FieldName\\":\\"submit\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": [{ \\\\n    \\\\\\"type\\\\\\": \\\\\\"guard\\\\\\",\\\\n    \\\\\\"pk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"sk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n},{\\\\n    \\\\\\"type\\\\\\": \\\\\\"add\\\\\\",\\\\n    \\\\\\"note\\\\\\":\\\\\\"!email\\\\\\"\\\\n},{ \\\\n    \\\\\\"type\\\\\\": \\\\\\"function\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\",\\\\n    \\\\\\"userpoolId\\\\\\":\\\\\\"\${UserPoolId}\\\\\\"\\\\n},{ \\\\n    \\\\\\"type\\\\\\": \\\\\\"db\\\\\\",\\\\n    \\\\\\"action\\\\\\": \\\\\\"create\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n}]\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverMutationcheck\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Mutation\\",\\"FieldName\\":\\"check\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": { \\\\n    \\\\\\"type\\\\\\": \\\\\\"guard\\\\\\",\\\\n    \\\\\\"pk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"sk\\\\\\": \\\\\\"123\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverMutationadd\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Mutation\\",\\"FieldName\\":\\"add\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": {\\\\n    \\\\\\"type\\\\\\": \\\\\\"add\\\\\\",\\\\n    \\\\\\"note\\\\\\":\\\\\\"!email\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverMutationhat\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Mutation\\",\\"FieldName\\":\\"hat\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": { \\\\n    \\\\\\"type\\\\\\": \\\\\\"function\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\",\\\\n    \\\\\\"userpoolId\\\\\\":\\\\\\"\${UserPoolId}\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}},\\"ResolverMutationnote\\":{\\"Type\\":\\"AWS::AppSync::Resolver\\",\\"DependsOn\\":\\"GraphQlSchema\\",\\"Properties\\":{\\"ApiId\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"ApiId\\"]},\\"TypeName\\":\\"Mutation\\",\\"FieldName\\":\\"note\\",\\"RequestMappingTemplate\\":{\\"Fn::Sub\\":[\\"\\\\n        #set( $myMap = {\\\\n            \\\\\\"ctx\\\\\\": $ctx,\\\\n            \\\\\\"action\\\\\\": { \\\\n    \\\\\\"type\\\\\\": \\\\\\"db\\\\\\",\\\\n    \\\\\\"action\\\\\\": \\\\\\"create\\\\\\",\\\\n    \\\\\\"table\\\\\\": \\\\\\"hi-dev\\\\\\"\\\\n}\\\\n        })\\\\n        \\\\n        { \\\\n            \\\\\\"version\\\\\\": \\\\\\"2017-02-28\\\\\\",\\\\n            \\\\\\"operation\\\\\\": \\\\\\"Invoke\\\\\\",\\\\n            \\\\\\"payload\\\\\\": $util.toJson($myMap)\\\\n        }\\",{\\"UserPoolId\\":{\\"Ref\\":\\"CognitoUserPoolMyUserPool\\"}}]},\\"ResponseMappingTemplate\\":\\"$util.toJson($ctx.result)\\",\\"Kind\\":\\"UNIT\\",\\"DataSourceName\\":{\\"Fn::GetAtt\\":[\\"GraphQlDslambdadatasource\\",\\"Name\\"]}}}},\\"Outputs\\":{\\"EventBridgeArn\\":{\\"Description\\":\\"EventBridgeArn\\",\\"Value\\":{\\"Fn::GetAtt\\":[\\"EventBridge\\",\\"Arn\\"]}},\\"ApiUrl\\":{\\"Description\\":\\"URL\\",\\"Value\\":{\\"Fn::GetAtt\\":[\\"GraphQlApi\\",\\"GraphQLUrl\\"]}},\\"ApiKey\\":{\\"Description\\":\\"ApiKey\\",\\"Value\\":{\\"Fn::GetAtt\\":[\\"GraphQlApiKeyDefault\\",\\"ApiKey\\"]}}}}"`;
